flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -576,7 +583,22 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
 										request.advanceToEndOfTime);
 
 									coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
-
+									// It is possible that the tasks has finished checkpointing at this point.
+									// So we need to complete this pending checkpoint.
```

Comment:
In an essence, this is a race condition between sending out snapshot state calls, and receiving ack's. And the race condition is caused by the fact, that `CheckpointCoordinator#receiveAcknowledgeMessage` is executed in the `ioExecutor`, while this logic here is happening on the timer thread?

If so, can we write this as a comment here?

Secondly, @ifndef-SleePy, would this problem go away with the last step of your refactor? I guess the idea should be, that both of those things would be executed in job manager's main thread executor, right?

Reply:
Ok, but at least `synchronized (lock) {` wouldn't be needed anymore, right?